### PR TITLE
Set state type in copy constructors

### PR DESCRIPTION
**Pull request by domire8**
## Description

- #20 

After back and forth in #24 and #31, I'm using the `set_type` in the relevant copy constructors to set the correct state type. There are also some good changes from #24 that I'm getting back.

One problem that this PR introduces:
The JSON encoding of any state now has `type: STATE`. I'm not quite sure what is going on in the background there so maybe @eeberhard can judge if this is problematic or not.

## Supporting information
The new tests of the refactor/operators branch now pass successfully. (However, other tests now fail with the operators refactoring)

## Review guidelines

Estimated Time of Review: 15 minutes
Review by commit

### CODE REVIEW

**Comment by eeberhard** [comment]
> Maybe that's a stupid suggestions, but could we even deprecate the State type in the `state.proto`? Isn't it the job of the encoders and decoders to create the correct object and the type was only there for information?

Not a stupid suggestion and it deserves some thinking. Initially the state type was mutable and represented families of state (e.g. CARTESIAN_STATE was used for multiple types). Then it was refactored to be one-to-one with the class type. Indeed the en/decoders now should not have to worry about the state type since the message type already represents the class, and the class constructors already define the type.

For now please just patch the protocol/python test that compares the JSON and I'll make a separate issue for it.

File: source/state_representation/include/state_representation/parameters/Parameter.hpp [hunk 1]
```diff
@@ -84,8 +84,11 @@ class Parameter : public ParameterInterface {
 
 template<typename T>
 template<typename U>
-Parameter<T>::Parameter(const Parameter<U>& parameter) :
-    Parameter<T>(parameter.get_name(), static_cast<T>(parameter.get_value())) {}
+Parameter<T>::Parameter(const Parameter<U>& parameter) :    Parameter<T>(parameter.get_name()) {
```
Comment (eeberhard) [inline source/state_representation/include/state_representation/parameters/Parameter.hpp:87]: whitespace formatting

**Review by eeberhard — COMMENTED** [review]
Ok, looks good and in my opinion is the most sensible resolution/compromise to the back and forth discussion we had

**Comment by domire8** [comment]
Ready for review again

**Review by eeberhard — APPROVED** [review]

**Pull request merged**